src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -215,27 +215,33 @@ internal sealed class RSession : IRSession, IRCallbacks {
         }
 
         public async Task EnsureHostStartedAsync(RHostStartupInfo startupInfo, IRSessionCallback callback, int timeout = 3000, CancellationToken cancellationToken = default(CancellationToken)) {
-            using (_disposeToken.Link(ref cancellationToken))
-            using (await _initializationLock.WaitAsync(cancellationToken)) {
-                if (_initializationTcs.Task.Status != TaskStatus.RanToCompletion || !_isHostRunning) {
-                    await StartHostAsyncBackground(startupInfo, callback, timeout, cancellationToken);
+            using (_disposeToken.Link(ref cancellationToken)) {
+                await TaskUtilities.SwitchToBackgroundThread();
+
+                using (await _initializationLock.WaitAsync(cancellationToken)) {
+                    if (_initializationTcs.Task.Status != TaskStatus.RanToCompletion || !_isHostRunning) {
```

Comment:
Fault or canceled mean that it has failed to start host last time, so we allow to try again